### PR TITLE
feat(hardened): wire hardened image recommendations into analysis pipeline

### PR DESCRIPTION
**Pull request by a-oren**
## Summary
- Add Camel multicast branch for hardened image recommendations with ProviderRoutePolicy telemetry
- Move OCI PURL resolution logic into `HardenedImageProvider.lookupBySbomId()` for proper class responsibility separation
- Add `@Inject` to `HardenedImageProvider` constructor to fix CDI bean resolution with multiple constructors
- Add integration tests (4) and unit tests (4) covering match/no-match/recommend-false/metrics scenarios

## Test plan
- [x] `HardenedImageIntegrationTest` — 4 WireMock-based `@QuarkusTest` integration tests
- [x] `HardenedImageProviderTest` — 4 new unit tests for `lookupBySbomId` edge cases
- [x] Full test suite passes (294 tests, 0 failures)

Closes TC-4809

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Wire hardened image recommendations into the analysis pipeline and expose them via batch analysis responses.

New Features:
- Add SBOM-based hardened image lookup in HardenedImageProvider using OCI PURLs to resolve base image references.
- Introduce a dedicated Camel route for hardened image recommendations and integrate it into the analysis multicast flow.
- Surface hardened image recommendations in analysis results with provider metrics recorded via ProviderRoutePolicy.

Enhancements:
- Annotate HardenedImageProvider with @Inject constructor injection to ensure proper CDI wiring.
- Refine hardened image index usage by mapping OCI PURLs to Docker-style base image references for recommendation lookup.

Tests:
- Add HardenedImageProvider unit tests covering OCI PURL lookup edge cases and non-OCI scenarios.
- Add WireMock-based HardenedImageIntegrationTest to verify hardened recommendations, recommend=false behavior, unmatched images, and metrics emission.

### CODE REVIEW

File: src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java [hunk 1]
```diff
@@ -126,6 +132,58 @@ public IndexedRecommendation lookup(String baseImageRef) {
...
+  public Map<PackageRef, IndexedRecommendation> lookupBySbomId(String sbomId) {
+    if (sbomId == null) {
+      return Collections.emptyMap();
+    }
+
+    var pkgRef = new PackageRef(sbomId);
+    if (!OCI_PURL_TYPE.equals(pkgRef.purl().getType())) {
+      return Collections.emptyMap();
+    }
+
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java:157]: [sdlc-workflow/verify-pr] Classified as: **code change request**

File: src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java [hunk 2]
```diff
@@ -126,6 +132,58 @@ public IndexedRecommendation lookup(String baseImageRef) {
...
+  private String buildDockerRef(PackageRef pkgRef) {
+    var qualifiers = pkgRef.purl().getQualifiers();
+    if (qualifiers == null) {
+      return null;
+    }
+    String repoUrl = qualifiers.get("repository_url");
+    if (repoUrl == null || repoUrl.isBlank()) {
+      return null;
+    }
+    String tag = qualifiers.get("tag");
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java:185]: [sdlc-workflow/verify-pr] Classified as: **question**

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java [hunk 3]
```diff
@@ -353,4 +353,58 @@ void testLockReleaseFailureDoesNotPropagate() throws Exception {
...
+  /// Verifies that lookupBySbomId returns the recommendation for a matching OCI PURL.
+  @Test
+  void testLookupBySbomIdMatchingOciPurl() {
+    // Given an index populated with a hardened nginx recommendation
+    IndexedRecommendation rec =
+        IndexedRecommendation.builder()
+            .packageName(new PackageRef(HARDENED_NGINX_PURL))
+            .vulnerabilities(Collections.emptyMap())
+            .sourceName("hardened")
+            .build();
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java]: [sdlc-workflow/verify-pr] Classified as: **suggestion**

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+  @Test
+  void testHardenedRecommendationIncludedInResponse() throws Exception {
+    // Given an OCI SBOM that matches a hardened recommendation in the index
+
+    // When sending a batch analysis with the matching OCI SBOM
+    String batchBody = createOciBatchRequest(OCI_SBOM_ID);
+    Map<String, AnalysisReport> response = sendBatchAnalysis(batchBody);
+
+    // Then the response contains the hardened recommendation
+    assertNotNull(response, "Response should not be null");
```
[the remark this answers is not in the document]
Comment (a-oren) [reply src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageIntegrationTest.java:169]: [sdlc-workflow/verify-pr] Classified as: **suggestion**

**Comment by a-oren** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai[bot] review

Classified as:
- review-body-4544288345-1 (defensively handle malformed PURL): **suggestion**
- review-body-4544288345-2 (use injected ObjectMapper in test): **suggestion**

File: src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (a-oren) [reply src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java:157]: [sdlc-workflow/verify-pr] Classified as **code change request** — sub-task [TC-4893](https://redhat.atlassian.net/browse/TC-4893) created to address this feedback.

File: src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (a-oren) [reply src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java:185]: [sdlc-workflow/verify-pr] Classified as **question** — this asks for clarification about digest vs tag handling; no code change needed. No sub-task created.

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (a-oren) [reply src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java]: [sdlc-workflow/verify-pr] Classified as **suggestion** — this proposes additional edge case tests that are not required by a documented convention. No sub-task created.

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageIntegrationTest.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (a-oren) [reply src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageIntegrationTest.java:169]: [sdlc-workflow/verify-pr] Classified as **suggestion** — this proposes strengthening the metrics test assertion beyond what CONVENTIONS.md requires. No sub-task created.

**Comment by a-oren** [comment]
[sdlc-workflow/verify-pr] Re: @sourcery-ai[bot] review

- **review-body-4544288345-1** (defensively handle malformed PURL): Classified as **code change request** (upgraded from suggestion) — this matches project convention: `HardenedImageResponseHandler.parseAndInvertMapping` demonstrates defensive PURL parsing with try-catch, and `testParseSkipsInvalidEntries` validates this pattern. Sub-task [TC-4893](https://redhat.atlassian.net/browse/TC-4893) created to address this feedback.
- **review-body-4544288345-2** (use injected ObjectMapper): Classified as **code change request** (upgraded from suggestion) — in `@QuarkusTest` classes, CDI-injected `ObjectMapper` should be used to match production configuration, consistent with `TrustifyIntegration`'s `@Inject ObjectMapper`. Sub-task [TC-4894](https://redhat.atlassian.net/browse/TC-4894) created to address this feedback.

[TC-4893]: https://redhat.atlassian.net/browse/TC-4893?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[TC-4894]: https://redhat.atlassian.net/browse/TC-4894?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by a-oren** [comment]
## Verification Report for TC-4809 (commit 371df78)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | WARN | 3 code change requests (1 original + 2 upgraded from suggestions); sub-tasks TC-4893, TC-4894 created |
| Root-Cause Investigation | DONE | 2 root-cause tasks created: TC-4895 (skill gap: defensive parsing), TC-4896 (convention gap: ObjectMapper) |
| Scope Containment | WARN | PackageItem.java listed in task but not modified; HardenedImageProvider.java modified but not listed |
| Diff Size | PASS | ~489 additions across 6 files, proportionate to task scope |
| Commit Traceability | PASS | Single commit references TC-4809 |
| Sensitive Patterns | PASS | No secrets or credentials detected |
| CI Status | PASS | All 3 CI checks pass (Integration Tests, Sourcery review, commitlint) |
| Acceptance Criteria | PASS | All 5 acceptance criteria satisfied |
| Test Quality | PASS | No repetitive tests, all tests documented, Eval Quality: N/A |
| Test Change Classification | ADDITIVE | All test files are new additions (no existing tests modified or removed) |
| Verification Commands | PASS | CI ran full test suite including new WireMock tests |

### Overall: WARN

**Issues requiring attention:**
- **TC-4893** (Blocks TC-4809): Add defensive PURL parsing in `HardenedImageProvider.lookupBySbomId` — wrap `new PackageRef(sbomId)` in try-catch to handle malformed PURL strings gracefully
- **TC-4894** (Blocks TC-4809): Use CDI-injected `ObjectMapper` in `HardenedImageIntegrationTest` instead of standalone `new ObjectMapper()`

**Root-cause tasks:**
- **TC-4895** (Related to TC-4809): Improve implement-task guidance for defensive parsing of untrusted input
- **TC-4896** (Related to TC-4809): Document ObjectMapper injection convention for `@QuarkusTest` classes in CONVENTIONS.md

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.11.0.*

**Comment by a-oren** [comment]
## Verification Report for TC-4809 (commit 7554510)

| Check | Result | Details |
|-------|--------|---------|
| Review Feedback | WARN | 3 code change requests (1 original + 2 upgraded); sub-tasks TC-4893 (Closed), TC-4894 (Closed) |
| Root-Cause Investigation | N/A | No new sub-tasks created in this run |
| Scope Containment | WARN | All task-specified files present; 12 additional out-of-scope files from base commit (TC-4808 infrastructure) |
| Diff Size | WARN | ~443 lines across 6 in-scope files; total PR includes ~1,412 lines across 17 files due to base commit |
| Commit Traceability | WARN | 2 of 3 commits reference TC-4809; others reference sub-tasks TC-4893, TC-4894 |
| Sensitive Patterns | PASS | No secrets or credentials detected |
| CI Status | PASS | All 3 checks pass (Integration Tests, Sourcery, commitlint) |
| Acceptance Criteria | PASS | 5 of 5 criteria met |
| Test Quality | WARN | Test Documentation: PASS; Repetitive Test Detection: WARN (4 lookupBySbomId empty-result tests could be parameterized); Eval Quality: N/A |
| Test Change Classification | ADDITIVE | All test files are new to this branch |
| Verification Commands | N/A | No verification commands specified |

### Overall: WARN

**Summary:** All acceptance criteria are met and CI passes. The WARN verdicts are informational:
- **Scope/Diff Size:** The PR includes infrastructure from the base commit (TC-4808/PR #627) which adds the hardened image provider foundation. The in-scope changes for TC-4809 are proportionate.
- **Commit Traceability:** Commits 5cce497 and 7554510 reference sub-tasks TC-4893 and TC-4894 respectively, which are review feedback fixes for this PR — this is expected behavior for the implement-task workflow.
- **Test Quality:** Four `lookupBySbomId` tests that assert empty results could be consolidated into a `@ParameterizedTest`, but this is a minor style improvement.

All review feedback sub-tasks (TC-4893, TC-4894) have been implemented and closed. The defensive PURL parsing and CDI-injected ObjectMapper fixes are included in this commit.

---
*This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.11.0.*

File: src/main/java/io/github/guacsec/trustifyda/integration/Constants.java [hunk 5]
```diff
@@ -87,6 +87,7 @@ private Constants() {}
   public static final String CACHE_LICENSES_HITS_PROPERTY = "cacheLicensesHits";
 
   public static final String TRUSTIFY_RECOMMEND_PATH = "/api/v2/purl/recommend";
+  public static final String TRUSTIFY_HARDENED_RECOMMEND_ROUTE = "hardenedRecommendations";
```
Comment (ruromero) [inline src/main/java/io/github/guacsec/trustifyda/integration/Constants.java]: This is not an existing pattern. Route names are not part of the Constants.

File: src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java [hunk 6]
```diff
@@ -116,6 +120,8 @@ void refresh() {
     }
   }
 
+  private static final String OCI_PURL_TYPE = "oci";
```
Comment (ruromero) [inline src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java]: constants should be declared before methods

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java [hunk 7]
```diff
@@ -353,4 +353,64 @@ void testLockReleaseFailureDoesNotPropagate() throws Exception {
     assertEquals(1, spyProvider.getIndex().size());
     assertNotNull(spyProvider.lookup("base:1.0"));
   }
+
+  /// Verifies that lookupBySbomId returns the recommendation for a matching OCI PURL.
```
Comment (ruromero) [inline src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java]: Either use `//` or `/**` as we do in the rest of the codebase

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java [hunk 8]
```diff
@@ -353,4 +353,64 @@ void testLockReleaseFailureDoesNotPropagate() throws Exception {
...
+    var result = provider.lookupBySbomId(ociPurl);
+
+    // Then the recommendation is returned
+    assertEquals(1, result.size());
+    var entry = result.entrySet().iterator().next();
+    assertEquals(normalized(HARDENED_NGINX_PURL), entry.getValue().packageName().ref());
+    assertEquals("hardened", entry.getValue().sourceName());
+  }
+
+  /// Verifies that lookupBySbomId returns empty for a non-OCI PURL.
```
Comment (ruromero) [inline src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java]: Same. As this `///` seems to be a Java23+ experimental feature we're not using it anywhere in the code

**Review by ruromero — CHANGES_REQUESTED** [review]
Some minor comments.
I will create a task to drop UBI recommendations as they no longer make sense after adding the Hardened Images.

File: src/main/java/io/github/guacsec/trustifyda/integration/Constants.java (hunk 5 above)
Comment (a-oren) [reply src/main/java/io/github/guacsec/trustifyda/integration/Constants.java]: [sdlc-workflow/verify-pr] Classified as **code change request**. Status: **Fixed** in commit d3114f4.

File: src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java (hunk 6 above)
Comment (a-oren) [reply src/main/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProvider.java]: [sdlc-workflow/verify-pr] Classified as **code change request**. Status: **Fixed** in commit d3114f4.

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java (hunk 7 above)
Comment (a-oren) [reply src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java]: [sdlc-workflow/verify-pr] Classified as **code change request**. Status: **Fixed** in commit d3114f4.

File: src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java (hunk 8 above)
Comment (a-oren) [reply src/test/java/io/github/guacsec/trustifyda/integration/providers/trustify/hardened/HardenedImageProviderTest.java]: [sdlc-workflow/verify-pr] Classified as **code change request**. Status: **Fixed** in commit d3114f4.

**Comment by a-oren** [comment]
[sdlc-workflow/verify-pr] Re: @ruromero review — Classified as **nit**. No action required.

**Review by a-oren — COMMENTED** [review]
## Verification Report — TC-4809

**Task:** Wire hardened image recommendations into analysis pipeline
**PR:** #631 | **Branch:** `TC-4809` | **HEAD:** `d3114f4`

### Summary

| Check | Verdict | Summary |
|---|---|---|
| Scope Containment | WARN | PR includes TC-4808 prerequisite; Constants.java and PackageItem.java not modified |
| Diff Size | WARN | 1,443 lines across 16 files (inflated by bundled TC-4808 prerequisite) |
| Commit Traceability | WARN | 1/4 commits references TC-4809; others reference TC-4893, TC-4894, or none |
| Sensitive Patterns | PASS | No hardcoded secrets or credentials found |
| CI Status | PASS | All 3 CI checks pass (Integration Tests, Sourcery, commitlint) |
| Acceptance Criteria | PASS | All 5 criteria satisfied |
| Verification Commands | N/A | CI covers `mvn test`; manual verification not run |
| Test Quality | WARN | Minor: repeated test setup patterns; `//` comments vs `/** */` inconsistency |
| Test Change Classification | ADDITIVE | Both test files are new additions |

### Review Feedback Status

| Thread | Reviewer | Classification | Status |
|---|---|---|---|
| Constants.java route constant | @ruromero | code change request | Fixed (d3114f4) |
| HardenedImageProvider constant placement | @ruromero | code change request | Fixed (d3114f4) |
| HardenedImageProviderTest `///` comments (×2) | @ruromero | code change request | Fixed (d3114f4) |
| Review body | @ruromero | nit | No action required |
| Test edge cases suggestion | sourcery-ai | suggestion | Kept as suggestion |
| Metrics assertion suggestion | sourcery-ai | suggestion | Kept as suggestion |

### Details

**Scope Containment (WARN):** The PR bundles TC-4808 (Hummingbird provider) with TC-4809 (pipeline wiring). The task spec listed `Constants.java` and `PackageItem.java` as files to modify — `Constants.java` was intentionally not modified per reviewer feedback (route constant inlined), and `PackageItem.java` was not needed (hardened recommendations use a different data model path via `IndexedRecommendation`).

**Acceptance Criteria (PASS):** All 5 criteria confirmed:
1. Hardened recommendation branch runs in parallel via Camel multicast (`TrustifyIntegration.java:191-196`)
2. Results embedded with `sourceName = "hardened"` in ProviderReport
3. ProviderRoutePolicy telemetry attached to `trustify-hardened-recommend` route
4. `recommend=false` skips hardened recommendations via choice/otherwise pattern
5. WireMock integration test validates 4 scenarios end-to-end

**No sub-tasks created.** All reviewer feedback has been addressed.

---
This comment was AI-generated by [sdlc-workflow/verify-pr](https://github.com/mrizzi/sdlc-plugins) v0.11.0.

**Review by ruromero — APPROVED** [review]
LGTM
